sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ChecksumCheckpointFileManager.scala

```diff
@@ -0,0 +1,513 @@
+        // Then previously created files won't have checksum.
+        case _: FileNotFoundException =>
+          logWarning(log"No checksum file found for ${MDC(PATH, path)}, " +
+            log"hence no checksum verification.")
+          None
+      }
+    }(threadPool)
+
+    val mainInputStreamFuture = Future {
+      underlyingFileMgr.open(path)
+    }(threadPool)
+
+    val mainStream = awaitResult(mainInputStreamFuture, Duration.Inf)
+    val checksumStream = awaitResult(checksumInputStreamFuture, Duration.Inf)
+
+    checksumStream.map { chkStream =>
+      new ChecksumFSDataInputStream(mainStream, path, chkStream, threadPool)
+    }.getOrElse(mainStream)
+  }
+
+  override def exists(path: Path): Boolean = underlyingFileMgr.exists(path)
+
+  override def delete(path: Path): Unit = {
+    // Allowing directly deleting the checksum file for orphan checksum file scenario
+    if (isChecksumFile(path)) {
+      deleteChecksumFile(path)
+    } else if (allowConcurrentDelete) {
+      // Ideally, we should first try to delete the checksum file
+      // before the main file, to avoid a situation where the main file is deleted but the
+      // checksum file deletion failed. The client might not call delete again if the main file
+      // no longer exists.
+      // But if allowConcurrentDelete is enabled, then we can do it concurrently for perf.
+      // But the client would be responsible for cleaning up potential orphan checksum files
+      // if it happens.
+      val checksumInputStreamFuture = Future {
+        deleteChecksumFile(getChecksumPath(path))
+      }(threadPool)
+
+      val mainInputStreamFuture = Future {
+        underlyingFileMgr.delete(path)
+      }(threadPool)
+
+      awaitResult(mainInputStreamFuture, Duration.Inf)
+      awaitResult(checksumInputStreamFuture, Duration.Inf)
+    } else {
+      // First delete the checksum file, then main file
+      deleteChecksumFile(getChecksumPath(path))
+      underlyingFileMgr.delete(path)
+    }
+  }
+
+  private def deleteChecksumFile(checksumPath: Path): Unit = {
+    try {
+      underlyingFileMgr.delete(checksumPath)
+      logDebug(log"Deleted checksum file ${MDC(PATH, checksumPath)}")
+    } catch {
+      case _: FileNotFoundException =>
+        // Ignore if file has already been deleted
+        // or the main file was created initially without checksum
+        logWarning(log"Skipping deletion of checksum file ${MDC(PATH, checksumPath)} " +
+          log"since it does not exist.")
+    }
+  }
+
+  override def isLocal: Boolean = underlyingFileMgr.isLocal
+
+  override def createCheckpointDirectory(): Path = {
+    underlyingFileMgr.createCheckpointDirectory()
+  }
+
+  override def close(): Unit = {
+    threadPool.shutdown()
+    // Wait a bit for it to finish up in case there is any ongoing work
+    // Can consider making this timeout configurable, if needed
+    val timeoutMs = 100
+    if (!threadPool.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS)) {
```

Comment:
Should we wait a bit longer here ? also - should we force the shutdown on some timeout ?

Reply:
done